py-polars/polars/utils/udfs.py

```diff
@@ -85,6 +109,80 @@ def _op(opname: str, argrepr: str, argval: Any) -> str:
         )
 
 
+def _ops_to_stack_value(ops: list[ByteCodeInfo], apply_target: str) -> StackValue:
+    """Take postfix bytecode and convert to an intermediate natural-order stack."""
+    if apply_target == "expr":
+        stack: list[StackValue] = []
+        for op in ops:
+            stack.append(
+                op[1]  # type: ignore[arg-type]
+                if op[0] in ("LOAD_FAST", "LOAD_CONST")
+                else (
+                    (stack.pop(), _op(*op))
+                    if op[0].startswith("UNARY_")
+                    else (stack.pop(-2), _op(*op), stack.pop(-1))
+                )
+            )
+
+        return stack[0]
+
+    # TODO: frame apply (account for BINARY_SUBSCR)
+    # TODO: series apply (rewrite col expr as series)
+    raise NotImplementedError(f"TODO: {apply_target!r} apply")
+
+
+def _bytecode_to_expression(
+    bytecode: list[ByteCodeInfo], col: str, apply_target: str, param_name: str
+) -> str | None:
+    """Take postfix bytecode and translate to polars expression string."""
+    # decompose bytecode into logical 'and'/'or' expression blocks (if present)
+    logical_blocks, logical_ops = defaultdict(list), []
+    jump_offset = 0
+    for idx, op in enumerate(bytecode):
+        if op[0] in _LOGICAL_OPCODES:
+            jump_offset = bytecode[idx + 1][3]
+            logical_ops.append(op)
+        else:
+            logical_blocks[jump_offset].append(op)
```

Comment:
right, so you "split" the sequence between those that happen before an `and/or`, and those which happen after? (just talking to myself here to help process it 😄 )

Reply:
Yup - but importantly it also associates each block with a particular jump target.